conveyor/slot.go

```diff
@@ -0,0 +1,238 @@
+/*
+ *    Copyright 2019 Insolar Technologies
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+package conveyor
+
+import (
+	"github.com/insolar/insolar/conveyor/queue"
+	"github.com/insolar/insolar/core"
+)
+
+// PulseState is the states of pulse inside slot
+type PulseState int
+
+//go:generate stringer -type=PulseState
+const (
+	Unallocated = PulseState(iota)
+	Future
+	Present
+	Past
+	Antique
+)
+
+// SlotState shows slot working mode
+type SlotState int
+
+//go:generate stringer -type=SlotState
+const (
+	Initializing = SlotState(iota)
+	Working
+	Suspending
+)
+
+const slotSize = 10000
+const slotElementDelta = 1000000
+
+// SlotDetails provides information about slot
+type SlotDetails interface {
+	getPulseNumber() core.PulseNumber
+	getNodeId() uint32
+	getPulseData() *core.Pulse
+	getNodeData() interface{}
+}
+
+// HandlersConfiguration contains configuration of handlers for specific pulse state
+// TODO: logic will be provided after pulse change mechanism
+type HandlersConfiguration struct {
+	state SlotState
+}
+
+// TODO: logic will be provided after pulse change mechanism
+func (s *HandlersConfiguration) getMachineConfiguration(smType int) StateMachineType {
+	return nil
+}
+
+// ActivationStatus represents status of work for slot element
+type ActivationStatus int
+
+//go:generate stringer -type=ActivationStatus
+const (
+	EmptyElement = ActivationStatus(iota)
+	ActiveElement
+	NotActiveElement
+)
+
+// ElementList is a list of slotElements with pointers to head and tail
+type ElementList struct {
+	head *slotElement
+	tail *slotElement
+}
+
+// popElement gets element from linked list (and remove it from list)
+func (l *ElementList) popElement() *slotElement {
+	result := l.head
+	if result == nil {
+		return nil
+	}
+	l.head = l.head.listNext
+	return result
+}
+
+// pushElement adds element to linked list
+func (l *ElementList) pushElement(element *slotElement) {
+	if l.head == nil {
+		l.head = element
+		l.tail = element
+	} else {
+		l.tail.listNext = element
+		l.tail = element
+	}
+}
+
+func (l *ElementList) len() int {
```

Comment:
it may be called multiple times and now it works for O(n). It's too long. Make it for O(1) ( for example,  add one more field to ElementList and adjust every push\pop\any other changing operation )

Reply:
this method was added just for easy testing, so I moved it to _test.go file. We are going to use push/pop very often, add changing length field in list make it longer. Worker or Conveyor itself don't need len info - they just try pop element as long as it exist